src/rust/plugin-registry/src/server/service.rs

```diff
@@ -129,32 +149,76 @@ impl PluginRegistryApi for PluginRegistry {
     #[tracing::instrument(skip(self, request), err)]
     async fn create_plugin(
         &self,
-        request: CreatePluginRequest,
+        request: futures::channel::mpsc::Receiver<CreatePluginRequest>,
     ) -> Result<CreatePluginResponse, Self::Error> {
-        let CreatePluginRequest {
+        let start_time = std::time::SystemTime::now();
+
+        let mut request = request;
+
+        let CreatePluginRequestMetadata {
             tenant_id,
-            plugin_artifact, // This could be huge, so don't clone it!
             display_name,
             plugin_type,
-        } = request;
-
-        self.ensure_artifact_size_limit(&plugin_artifact)?;
-
-        let plugin_id = generate_plugin_id(&tenant_id, plugin_artifact.as_slice());
+        } = match request.next().await {
+            Some(CreatePluginRequest::Metadata(m)) => m,
+            _ => {
+                return Err(Self::Error::StreamInputError(
+                    "Expected request 0 to be Metadata",
+                ));
+            }
+        };
 
+        let plugin_id = generate_plugin_id();
         let s3_key = generate_artifact_s3_key(plugin_type, &tenant_id, &plugin_id);
 
+        // Convert the incoming request stream of CreatePluginRequest::Chunk
+        // into a stream of Bytes to be sent to Rusoto S3.put_object.
+        let body_stream = request.then(|result| async {
+            match result {
+                CreatePluginRequest::Chunk(c) => Ok(Bytes::from(c)),
+                _ => Err(io_input_error("Expected request 1..N to be Chunk")),
+            }
+        });
```

Comment:
I think ideally we'd do this via S3 multipart uploads, that way we can drop the data from memory, but I would prefer that we do it this way for now and just add a ticket for that later.

Reply:
Rusoto automatically does that for us! You pass it a Stream.

Reply:
wow jk i think i must have made that up

Reply:
i changed it to use multipart